x/twap/logic_test.go

```diff
@@ -42,76 +43,287 @@ func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
+		},
+		"update block height": {
+			inputRecord:       newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateBlockHeight: true,
+			expRecord:         newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+		"update block time": {
+			inputRecord:     newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateSpotPrice: true,
+			updateBlockTime: true,
 			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
 		},
+		"spot price changed": {
+			inputRecord:     newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateSpotPrice: true,
+			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.SetupTest()
+			poolId := s.PrepareBalancerPoolWithCoins(defaultTwoAssetCoins...)
+
+			if test.updateBlockHeight {
+				s.Ctx = s.Ctx.WithBlockHeight(s.Ctx.BlockHeight() + 1)
+			}
+
+			if test.updateBlockTime {
+				s.Ctx = s.Ctx.WithBlockTime(s.Ctx.BlockTime().Add(time.Minute))
+			}
+
+			if test.updateSpotPrice {
+				s.RunBasicSwap(poolId)
+			}
+
+			newRecord := s.twapkeeper.UpdateRecord(s.Ctx, test.inputRecord)
+
+			spotPrice, err := s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, newRecord.Asset0Denom, newRecord.Asset1Denom)
+			s.Require().NoError(err)
+
+			s.Require().Equal(spotPrice, newRecord.P0LastSpotPrice)
+			s.Require().Equal(s.Ctx.BlockHeight(), newRecord.Height)
+			s.Require().Equal(s.Ctx.BlockTime(), newRecord.Time)
+			s.Require().Equal(poolId, newRecord.PoolId)
+
+			// we test specific math within deeper unit tests, test only for accumulator increase
+			if test.updateBlockTime {
+				s.Require().True(newRecord.P0ArithmeticTwapAccumulator.GT(zeroDec))
+				s.Require().True(newRecord.P1ArithmeticTwapAccumulator.GT(zeroDec))
+			} else {
+				s.Require().False(newRecord.P0ArithmeticTwapAccumulator.GT(zeroDec))
+				s.Require().False(newRecord.P1ArithmeticTwapAccumulator.GT(zeroDec))
+			}
+		})
+	}
+}
+
+// TestPruneRecords tests that all twap records earlier than
+// current block time - RecordHistoryKeepPeriod are pruned from the store.
+func (s *TestSuite) TestPruneRecords() {
+	recordHistoryKeepPeriod := s.twapkeeper.RecordHistoryKeepPeriod(s.Ctx)
+
+	tMin2Record, tMin1Record, baseRecord, tPlus1Record := s.createTestRecordsFromTime(baseTime.Add(-recordHistoryKeepPeriod))
+
+	// non-ascending insertion order.
+	recordsToPreSet := []types.TwapRecord{tPlus1Record, tMin1Record, baseRecord, tMin2Record}
+
+	expectedKeptRecords := []types.TwapRecord{baseRecord, tPlus1Record}
+	s.SetupTest()
+	s.preSetRecords(recordsToPreSet)
+
+	ctx := s.Ctx
+	twapKeeper := s.twapkeeper
+
+	ctx = ctx.WithBlockTime(baseTime)
+
+	err := twapKeeper.PruneRecords(ctx)
+	s.Require().NoError(err)
+
+	s.validateExpectedRecords(expectedKeptRecords)
+}
+
+func TestRecordWithUpdatedAccumulators(t *testing.T) {
```

Comment:
Is there a reason this test isn't part of the testSuite?

Reply:
This test case isn't using any methods from test suite, there were no reason it should be a part of test suite. Do you think it should go under test suite? If so, why do you think so?

Reply:
Yup, you're right, i do not have strong preference here. Donot want to force it anything, if its not being used.

Reply:
Things that don't use keeper methods shouldn't use the suite imo